worker/uniter/operation/executor.go

```diff
@@ -27,27 +27,52 @@ var (
 )
 
 type executor struct {
-	file               *StateFile
+	stateOps           *StateOps
 	state              *State
 	acquireMachineLock func(string) (func(), error)
 }
 
-// NewExecutor returns an Executor which takes its starting state from the
-// supplied path, and records state changes there. If no state file exists,
-// the executor's starting state will include a queued Install hook, for
-// the charm identified by the supplied func.
-func NewExecutor(stateFilePath string, initialState State, acquireLock func(string) (func(), error)) (Executor, error) {
-	file := NewStateFile(stateFilePath)
-	state, err := file.Read()
-	if err == ErrNoStateFile {
-		state = &initialState
+type ExecutorConfig struct {
+	StateOps      UniterStateOps
+	StateFilePath string
+	InitialState  State
+	AcquireLock   func(string) (func(), error)
+}
+
+func (e ExecutorConfig) validate() error {
+	if e.StateOps == nil {
+		return errors.NotValidf("executor config with nil state ops")
+	}
+	logger.Criticalf("validate queued step is %s", e.InitialState.Step)
+	// The uniter always starts with a Queued initial step.
+	//if e.InitialState.Step == Queued {
+	//	return errors.NotValidf("executor config with no initial state")
+	//}
+	return nil
+}
+
+// NewExecutor returns an Executor which takes its starting state from
+// the controller, and records state changes there. If no saved state
+// exists, the executor's starting state will be the supplied InitialState.
+func NewExecutor(cfg ExecutorConfig) (Executor, error) {
+	if err := cfg.validate(); err != nil {
+		return nil, err
+	}
+	// TODO: hml 03-18-2020
+	// handle fallback order better in relation to ugprades
+	// of a unit from state stateOps to state on controller
+	// or implement upgrade steps.
+	stateOps := NewStateOps(cfg.StateFilePath, cfg.StateOps)
```

Comment:
Since `cfg.StateOps` implements the same interface as the `StateOps` type, I think it would be much cleaner if the caller instantiated it directly (it is aware of the file path) and passed it as part of the configuration.

It would also make it easier to test the executor by mocking it.

Reply:
This will be rejiggered now that I can remove the StateFilePath part.